zerver/tests/test_tusd.py

```diff
@@ -0,0 +1,185 @@
+import os
+
+from django.conf import settings
+
+from zerver.lib.test_classes import ZulipTestCase
+from zerver.lib.test_helpers import (
+    create_s3_buckets,
+    use_s3_backend,
+)
+from zerver.lib.upload.s3 import upload_image_to_s3
+
+
+class ZephyrTest(ZulipTestCase):
+    def test_tusd_auth(self) -> None:
+        body = {
+            "Upload": {
+                "ID": "",
+            },
+        }
+        result = self.client_post(
+            f"/tusd/hooks?secret={settings.SHARED_SECRET}",
+            body,
+            content_type="application/json",
+            HTTP_HOOK_NAME="pre-create",
+        )
+        self.assert_json_error_contains(
+            result, "Not logged in: API authentication or user session required", 401
+        )
+
+    def test_tusd_pre_create_hook(self) -> None:
+        self.login("hamlet")
+        body = {
+            "Upload": {
+                "ID": "",
+                "IsFinal": False,
+                "IsPartial": False,
+                "MetaData": {
+                    "filename": "jupyterlab-galata-report.zip",
+                    "filetype": "application/zip",
+                    "name": "jupyterlab-galata-report.zip",
+                    "type": "application/zip",
+                },
+                "Offset": 0,
+                "PartialUploads": None,
+                "Size": settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024 - 100,
+                "SizeIsDeferred": False,
```

Comment:
What happens in our impolementation if the client does [defer the size](https://tus.io/protocols/resumable-upload#upload-defer-length)?  How can we make clients which do that not fail their uploads?

Reply:
There is no handler for this, I will take care of this and give a update here. I will assert that file size is not deferred.

Reply:
Asserted `SizeIsDeferred` to be false in pre-create hook.